stub/src/main/java/io/grpc/stub/StreamObservers.java

```diff
@@ -0,0 +1,92 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.stub;
+
+import com.google.common.base.Preconditions;
+
+import io.grpc.ExperimentalApi;
+
+import java.util.Iterator;
+
+/**
+ * Utility functions for working with {@link StreamObserver} and it's common subclasses like
+ * {@link CallStreamObserver}.
+ */
+@ExperimentalApi
+public class StreamObservers {
+  /**
+   * Copy the values of an {@link Iterator} to the target {@link StreamObserver}. If the
+   * target is an instance of {@link CallStreamObserver} then values will only be taken from
+   * the source if flow control will admit them.
+   *
+   * @param source of values expressed as an {@link Iterator}.
+   * @param target {@link StreamObserver} which accepts values from the source.
+   */
+  public static <V> void copy(final Iterator<V> source, StreamObserver<V> target) {
+    Preconditions.checkNotNull(source);
+    Preconditions.checkNotNull(target);
+    if (target instanceof CallStreamObserver) {
+      final CallStreamObserver<V> callStream = (CallStreamObserver<V>) target;
+      callStream.setOnReadyHandler(new Runnable() {
+        @Override
+        public void run() {
+          while (callStream.isReady() && source.hasNext()) {
+            callStream.onNext(source.next());
+          }
+          if (!source.hasNext()) {
+            callStream.onCompleted();
+          }
+        }
+      });
+    } else {
+      while (source.hasNext()) {
+        target.onNext(source.next());
```

Comment:
This is a bit scary because the user called this method for flow control but none will be taking place. I can see how that would be useful during tests though...

More importantly, `onNext()` can't be called at this point because `call.start()` hasn't been called.